sigmet/au3_functions.py

```diff
@@ -56,84 +57,63 @@ def standardize(series):
     a given element in the database
-
     Parameters
     ----------
     series : pd.Series
-        Find the starting date of a given series
-    start_date : pd.DateTime
-        The start_date at which to run my_min
-    end_date : pd.DateTime
-        The cutoff date for the series
+        Time series data, with datetime indices
+    user_start : pd.DateTime
+        Date after which to start searching for a recession, exclusive
+    user_end : pd.DateTime
+        Date before which to start searching for a recession, exclusive
     threshold : int
-        Threshold for my_min, called within this method
-
+        Threshold for local maxima comparison
     Returns
     -------
     pd.DateTime
-        The start date
-    """
-
-    # Differencing threshold
-    THRESHOLD = 1
-
-    # Get the most rcent date and filter series
-    last_date = series.index.sort_values(ascending=False).iloc[0]
-    series = series.index < end_date
-    series = series.ioc[1:]
-
-    # Initialize differences array for maxes
-    diffs = series.to_numpy()
-
-    # Populate is_max array using differencing
-    is_max = np.array([(
-        diffs[i] >= 0) and (
-            diffs[i + 1] <= 0) and (
-                diffs[i + 1] - diffs[i] <= THRESHOLD)
-                for i in range(len(series) - 1)])
-
-    is_max = np.append(is_max, False)
-
-    # If there are no local maxes, return the last date
-    if np.count_nonzero(is_max) == 0:
-        return last_date
-
-    maxes = is_max
-
-    # Get the proper minimum date
-    recession_minimum_date = my_min(series, start_date, end_date, threshold)
-
-    # Get max before that date
-    for i in range(len(series)):
-        if (maxes[i] == 1.0 & series.index[i] < recession_minimum_date):
-            max_index = i
-            break
-
-    # Get the actual start date
-    start_date = series.index[max_index]
-    return start_date
-
-
-def minus_helper(col, x):
-    """Subtracts two elements in a series
-
-    Parameters
-    ----------
-    col : pd.Series
-        The input column
-    x : int
-        The location
-
-    Returns
-    -------
-    int
-        The result
+        The start date of the largest detected recession in window. Returns most recent
+        date in window if no recession is found.
     """
-
-    return col.iloc[x] - col.iloc[x - 1]
+    assert(ma_window > 0, "Moving average window cannot be less than 1")
+    # filter series to window between start and end date, smooth moving average
+    # NOTE: we inclusively filter with user_start b/c when we run the differencing it will no longer be part of the series
+    filtered_series = series.rolling(ma_window).mean().loc[(series.index >= user_start) & (series.index < user_end)]
+    if filtered_series.hasnans:
+        raise ValueError("Moving average value too large for search window.")
+
+    # special case if monotonic decreasing, want to check for largest first derivative
+    if filtered_series.is_monotonic_decreasing:
```

Comment:
Lucas & Camille:

Why not just return the first value of the series? Just because we are getting the first value at which the trend seems to decrease faster, that won't always mean that the recession starts there.

Reply:
change to check if derivative change is > 50% of the current negative slope.